### DISCUSSION TITLE
git_remote_push() needs to handle push rejection

### DISCUSSION THREAD

**Discussion by anatol** [libgit2/libgit2#6832] [Q&A] [answered]
I am using libgit2 v1.8.1 and testing its functionality against the git CLI.

I am trying to push a branch to our remote repo and it fails:

```
$ git push  myprotocol://repo/users/anatol HEAD:refs/heads/__test
Enumerating objects: 3, done.
Counting objects: 100% (3/3), done.
Writing objects: 100% (3/3), 207 bytes | 207.00 KiB/s, done.
Total 3 (delta 0), reused 0 (delta 0), pack-reused 0 (from 0)
To myprotocol://repo/users/anatol
 ! [remote rejected] HEAD -> __test (Pushing to the ref 'refs/heads/__test' requires a justification: git push -o push-justification='bug_id')
error: failed to push some refs to 'myprotocol://repo/users/anatol'
```

The remote server rejected my push and expects me to specify `-o push-justification='bug_id'` option. It is expected.

Now I am trying to push with libgit2

```c
  char *refspec[] = {(char *)"HEAD:refs/heads/__test"};
  const git_strarray refspecs = {refspec, 1};
  int ret = git_remote_push(remote, &refspecs, nullptr);
```

The `git_remote_push` function above returns `0` as a sign of success.

But there are no any branch updates on the remotes. The remote rejected my push. I captured the traffic and I see a rejection message from the remote:

```
009b000eunpack ok
0084ng refs/heads/__test Pushing to the ref 'refs/heads/__test' requires a justification: git push -o push-justification='bug_id'
00000000
```

yet libgit2 happily ignores the error.

I would expect that `git_remote_push` returns a error code and the error message set to `Pushing to the ref 'refs/heads/__test' requires a justification: git push -o push-justification='bug_id'`.

**Comment by lrm29** [comment, marked as answer]
I can provide a solution for you: The callback you need to register is git_push_update_reference_cb: https://libgit2.org/libgit2/#HEAD/group/callback/git_push_update_reference_cb. status will contain the rejection message you see.

**Comment by ethomson** [comment]
@lrm29 is correct here — to add some color, the rationale for this behavior is that there can be multiple ref updates in a single push, each one can fail or succeed individually. So the reporting of what tips the remote has updated is independent of the overall success or failure of the operation.

**Comment by anatol** [comment]
Thank you for the quick reply! I'll look at it.

Handling errors from remote sounds like a common functionality. Are there any existing callback implementations that handle error codes?
